src/cascadia/TerminalAzBridge/ConsoleInputReader.cpp

```diff
@@ -50,15 +51,32 @@ std::optional<std::wstring_view> ConsoleInputReader::Read()
                         continue; // we've consumed it -- only dispatch it if we get a low
                     }
 
-                    if (IS_LOW_SURROGATE(keyEvent.uChar.UnicodeChar) && _highSurrogate)
+                    if (IS_LOW_SURROGATE(keyEvent.uChar.UnicodeChar))
                     {
-                        _convertedString.push_back(*_highSurrogate);
-                        _highSurrogate.reset();
+                        // No matter what we do, we want to destructively consume the high surrogate
+                        if (const auto oldHighSurrogate{ std::exchange(_highSurrogate, std::nullopt) })
+                        {
+                            _convertedString.push_back(*_highSurrogate);
+                        }
+                        else
+                        {
+                            // If we get a low without a high surrogate, we've done everything we can.
+                            // This is an illegal state.
+                            _convertedString.push_back(UNICODE_REPLACEMENT);

```

Comment:
what i have written is the same as `optional.value_or(REPLACEMENT)`, but with the added `continue`.

I figured it was more clear this way, and we get the added benefit of exchanging the optional with nullopt

Reply:
works for me.